Trace.py

```diff
@@ -64,20 +64,19 @@ def _dump_pfb(xbox):
     return bytes(buffer)
 
 
-def _read_pgraph_rdi(xbox, offset, count):
+def _read_pgraph_rdi(xbox: Xbox, offset: int, count: int) -> bytes:
     # FIXME: Assert pusher access is disabled
     # FIXME: Assert PGRAPH idle
 
     NV10_PGRAPH_RDI_INDEX = 0xFD400750
     NV10_PGRAPH_RDI_DATA = 0xFD400754
 
-    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, offset)
-    data = bytearray()
-    for _ in range(count):
-        word = xbox.read_u32(NV10_PGRAPH_RDI_DATA)
-        data += struct.pack("<L", word)
+    original_offset = ExchangeU32.exchange_u32(xbox, NV10_PGRAPH_RDI_INDEX, offset)
+
+    data = xbox.read(NV10_PGRAPH_RDI_DATA, count * 4)
```

Comment:
> that is mostly the overhead of XBDM

Yes, there's a lot of networking overhead.
Both, for small transactions which are done individually, but also for large transfers which are very slow.

I had always planned to (eventually) run nv2a-trace through https://github.com/JayFoxRox/xbe-loader, nkpatcher or trainers (by integrating nxdk-rdt support, or some plugin system).
There has even been work to emulate the Xbox NIC so we have a real NIC (for the debugger) and a virtual NIC (for Xbox games to use).

The protocol of XBDM isn't ideal and the high level operations in xboxpy (which extend XBDM) are not ideal either.
For nv2a-trace, a hotfix has been to write some custom assembly for the most critical portions. Ideally, we'd be able to load C plugins into a running XBE, and we'd be able to do so without XBDM.

So ideally, nv2a-trace would be partially a C/ASM program running on the Xbox, partially a Python script for capturing on the PC side, and then Python (or Javascript / Typescript, due to HTML being so useful for portable UIs) for display.

Also note that I also worked on dxt support for nxdk for a while (starting with https://github.com/JayFoxRox/xbox-fps-overlay using https://github.com/JayFoxRox/nxdk/pulls?q=is%3Apr+is%3Aopen+dxt), so that was also an option I considered (although I didn't like how crashy the dxt-loader in the XDK was, and wanted to avoid proprietary stuff like the XDK).

Anyhow, I digress.

Long story short: Yes, XBDM is slow. And while we have options to make it fast, I'd ideally like to see a solution that is FOSS and runs on retail Xboxes and also works with multiplayer games (which often break with XBDM).

Reply:
This makes absolute sense to me. I actually tried following in your footsteps with dxt's when I was putting together https://github.com/abaire/xbdm_gdb_bridge but the nxdk had changed to the point that I wasn't able to get it working well. The fact that a bad dxt can prevent the box from booting far enough to allow deletion of the dxt is also a huge concern for me.

Being able to patch in a C program would be huge and is something I'd already put on my wishlist to investigate (also the ability to ensure they're loaded past 64 megs for modded boxes to deal with scenarios where most of the memory is spoken for, but perhaps that's already handled by `xboxpy`?). While I really appreciate the continuous output of nv2a-trace, it does seem like it'd be a reasonable tradeoff if a capture could be done far faster on the xbox itself, storing dumps on the harddrive with separate programs built to process the dumps into nicer formats after the fact.